app/controllers/houseConsignment/index/items/additionalReference/AdditionalReferenceNumberController.scala

```diff
@@ -0,0 +1,116 @@
+          val form              = formProvider(viewModel.requiredError)
+          val preparedForm = request.userAnswers.get(AdditionalReferenceNumberPage(houseConsignmentIndex, itemIndex, additionalReferenceIndex)) match {
+            case None        => form
+            case Some(value) => form.fill(value)
+          }
+          Ok(view(preparedForm, request.userAnswers.mrn, viewModel, paragraphRequired))
+      }
+
+  def onSubmit(arrivalId: ArrivalId, mode: Mode, houseConsignmentIndex: Index, itemIndex: Index, additionalReferenceIndex: Index): Action[AnyContent] = actions
+    .requireData(arrivalId)
+    .andThen(getMandatoryPage(AdditionalReferencePage(houseConsignmentIndex, itemIndex, additionalReferenceIndex)))
+    .async {
+      implicit request =>
+        val paragraphRequired = isParagraphRequired(mode, houseConsignmentIndex, itemIndex, additionalReferenceIndex)
+        val viewModel         = viewModelProvider.apply(arrivalId, mode, houseConsignmentIndex, itemIndex, additionalReferenceIndex)
+        val form              = formProvider(viewModel.requiredError)
+        form
+          .bindFromRequest()
+          .fold(
+            formWithErrors => Future.successful(BadRequest(view(formWithErrors, request.userAnswers.mrn, viewModel, paragraphRequired))),
+            value => redirect(value, houseConsignmentIndex, itemIndex, additionalReferenceIndex, mode)
+          )
+    }
+
+  private def isParagraphRequired(mode: Mode, houseConsignmentIndex: Index, itemIndex: Index, additionalReferenceIndex: Index)(implicit
+    request: Request
+  ): Boolean =
+    mode match {
+      case CheckMode  => false
+      case NormalMode => otherAdditionalReferenceTypes(houseConsignmentIndex, itemIndex, additionalReferenceIndex).contains(request.arg)
+    }
+
+  private def otherAdditionalReferenceTypes(houseConsignmentIndex: Index, itemIndex: Index, additionalReferenceIndex: Index)(implicit
+    request: Request
+  ): Seq[AdditionalReferenceType] = {
+    val numberOfAdditionalReferences = request.userAnswers.get(AdditionalReferencesSection(houseConsignmentIndex, itemIndex)).length
+    (0 until numberOfAdditionalReferences)
+      .map(Index(_))
+      .filterNot(_ == additionalReferenceIndex)
+      .map(AdditionalReferencePage(houseConsignmentIndex, itemIndex, _))
+      .flatMap(request.userAnswers.get(_))
+  }
+
+  private def redirect(
+    value: String,
+    houseConsignmentIndex: Index,
+    itemIndex: Index,
+    additionalReferenceIndex: Index,
+    mode: Mode
+  )(implicit request: MandatoryDataRequest[_]): Future[Result] =
+    for {
+      updatedAnswers <- Future.fromTry(
+        request.userAnswers.set(AdditionalReferenceNumberPage(houseConsignmentIndex, itemIndex, additionalReferenceIndex), value)
+      )
+      _ <- sessionRepository.set(updatedAnswers)
+    } yield Redirect(navigator.nextPage(AdditionalReferenceNumberPage(houseConsignmentIndex, itemIndex, additionalReferenceIndex), mode, request.userAnswers))
+}
```

Comment:
In the Redirect(), whether we need to use `request.userAnswers` or `updatedAnswers`

Reply:
It's the request.userAnswers we use.